src/lib/utils/meta_tables/meta_system_utilization_table.cpp

```diff
@@ -0,0 +1,270 @@
+
+  return active_ns;
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+/*
+ * Returns the time in ns that THIS process has spent on the CPU
+ * since an arbitrary point in the past.
+*/
+uint64_t MetaSystemUtilizationTable::_get_process_cpu_time() {
+  // CLOCK_PROCESS_CPUTIME_ID:
+  // Per-process CPU-time clock (measures CPU time consumed by all threads in the process).
+#ifdef __linux__
+  struct timespec time_spec {};
+  [[maybe_unused]] const auto ret = clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &time_spec);
+  DebugAssert(ret == 0, "Failed in clock_gettime");
+
+  size_t cpu_count;
+  if (numa_available() != -1) {
+    cpu_count = numa_num_task_cpus();
+  } else {
+    cpu_count = _get_cpu_count();
+  }
+
+  const auto active_ns = (time_spec.tv_sec * std::nano::den + time_spec.tv_nsec) / cpu_count;
+
+  return active_ns;
+#endif
+
+#ifdef __APPLE__
+  const auto active_ns = clock_gettime_nsec_np(CLOCK_PROCESS_CPUTIME_ID);
+  DebugAssert(active_ns != 0, "Failed in clock_gettime_nsec_np");
+
+  return active_ns / _get_cpu_count();
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+/*
+ * Returns a struct that contains the available and free memory size in bytes.
+ * - Free memory is unallocated memory.
+ * - Available memory includes free memory and currently allocated memory that
+ *   could be made available (e.g. buffers, caches ...).
+ *   This is not equivalent to the total memory size, since certain data can not
+ *   be paged at any time.
+*/
+MetaSystemUtilizationTable::SystemMemoryUsage MetaSystemUtilizationTable::_get_system_memory_usage() {
+#ifdef __linux__
+  std::ifstream meminfo_file;
+  meminfo_file.open("/proc/meminfo", std::ifstream::in);
+  DebugAssert(meminfo_file.is_open(), "Unable to open /proc/meminfo");
+
+  MetaSystemUtilizationTable::SystemMemoryUsage memory_usage{};
+  std::string meminfo_line;
+  while (std::getline(meminfo_file, meminfo_line)) {
+    if (meminfo_line.rfind("MemFree", 0) == 0) {
+      memory_usage.free_memory = _parse_value_string(meminfo_line)[0] * 1024;
+    } else if (meminfo_line.rfind("MemAvailable", 0) == 0) {
+      memory_usage.available_memory = _parse_value_string(meminfo_line)[0] * 1024;
+    }
+  }
+  meminfo_file.close();
+
+  return memory_usage;
+#endif
+
+#ifdef __APPLE__
+  int64_t physical_memory;
+  size_t size = sizeof(physical_memory);
+  [[maybe_unused]] auto ret = sysctlbyname("hw.memsize", &physical_memory, &size, nullptr, 0);
+  DebugAssert(ret == 0, "Failed to call sysctl hw.memsize");
+
+  vm_size_t page_size;
+  vm_statistics64_data_t vm_statistics;
+  mach_msg_type_number_t count = sizeof(vm_statistics) / sizeof(natural_t);
+  ret = host_page_size(mach_host_self(), &page_size);
+  DebugAssert(ret == KERN_SUCCESS, "Failed to get page size");
+  ret = host_statistics64(mach_host_self(), HOST_VM_INFO, reinterpret_cast<host_info64_t>(&vm_statistics), &count);
+  DebugAssert(ret == KERN_SUCCESS, "Failed to get host_statistics64");

```

Comment:
Did you get this from somewhere or do you have any resources that you could link?

Reply:
https://stackoverflow.com/a/1911863
should that be documented in a code comment as well?

Reply:
Doesn't hurt.